APMrover2/sailboat.cpp

```diff
@@ -20,138 +20,212 @@ To Do List
+
+    // @Param: HEEL_MAX
+    // @DisplayName: Sailing maximum heel angle
+    // @Description: When in auto sail trim modes the heel will be limited to this value using PID control
+    // @Units: deg
+    // @Range: 0 90
+    // @Increment: 1
+    // @User: Standard
+    AP_GROUPINFO("HEEL_MAX", 5, Sailboat, sail_heel_angle_max, 15),
+
+    // @Param: SAIL_NO_GO_ANGLE
+    // @DisplayName: Sailing no go zone angle
+    // @Description: The typical closest angle to the wind the vehicle will sail at. the vehicle will sail at this angle when going upwind
+    // @Units: deg
+    // @Range: 0 90
+    // @Increment: 1
+    // @User: Standard
+    AP_GROUPINFO("NO_GO_ANGLE", 6, Sailboat, sail_no_go, 45),
+
+    AP_GROUPEND
+};
+
+
+/*
+  constructor
+ */
+Sailboat::Sailboat()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+
+    // sailboat defaults
+    if (enabled()) {
+        rover.g2.crash_angle.set_default(0);
+        rover.g2.loit_type.set_default(1);
+        rover.g2.loit_radius.set_default(5);
+        rover.g.waypoint_overshoot.set_default(10);
+    }
+}
+
 // update mainsail's desired angle based on wind speed and direction and desired speed (in m/s)
-void Rover::sailboat_update_mainsail(float desired_speed)
+void Sailboat::update_mainsail(float desired_speed)
 {
-    if (!g2.motors.has_sail()) {
+    if (!enabled()) {
         return;
     }
 
     // relax sail if desired speed is zero
     if (!is_positive(desired_speed)) {
-        g2.motors.set_mainsail(100.0f);
+        rover.g2.motors.set_mainsail(100.0f);
         return;
     }
 
     // + is wind over starboard side, - is wind over port side, but as the sails are sheeted the same on each side it makes no difference so take abs
-    float wind_dir_apparent = fabsf(g2.windvane.get_apparent_wind_direction_rad());
+    float wind_dir_apparent = fabsf(rover.g2.windvane.get_apparent_wind_direction_rad());
     wind_dir_apparent = degrees(wind_dir_apparent);
 
     // set the main sail to the ideal angle to the wind
-    float mainsail_angle = wind_dir_apparent - g2.sail_angle_ideal;
+    float mainsail_angle = wind_dir_apparent -sail_angle_ideal;
 
     // make sure between allowable range
-    mainsail_angle = constrain_float(mainsail_angle, g2.sail_angle_min, g2.sail_angle_max);
+    mainsail_angle = constrain_float(mainsail_angle,sail_angle_min, sail_angle_max);
 
     // linear interpolate mainsail value (0 to 100) from wind angle mainsail_angle
-    float mainsail = linear_interpolate(0.0f, 100.0f, mainsail_angle, g2.sail_angle_min, g2.sail_angle_max);
+    float mainsail = linear_interpolate(0.0f, 100.0f, mainsail_angle,sail_angle_min,sail_angle_max);
 
     // use PID controller to sheet out
-    const float pid_offset = g2.attitude_control.get_sail_out_from_heel(radians(g2.sail_heel_angle_max), G_Dt) * 100.0f;
+    const float pid_offset =rover.g2.attitude_control.get_sail_out_from_heel(radians(sail_heel_angle_max), rover.G_Dt) * 100.0f;
```

Comment:
it seems you remove a space by changing all g2. by rover.g2 on this file !

Reply:
fixed, thanks